File: py/data_structures/Deque.py
class DequeNode:
    """
    This class represents a node based off a doubly linked list node, meant to be used
    within the Deque data structure.
    """ 
    def __init__(self, val):
        self.val = val
        self.prev = None
        self.next = None


class Deque:
    """
    This class represents the deque data structure. 
    """ 
    def __init__(self):
        self.head = None
        self.tail = None
    
    def peek_first(self):
        if self.head:
            return self.head.val
        else:
            raise IndexError("peek first in empty deque")
    
    def peek_last(self):
        if self.tail:
            return self.tail.val
        else:
            raise IndexError("peek last in empty deque")

    def add_first(self, val):
        new_head = DequeNode(val)
        if self.head:
            self.head.prev = new_head
            new_head.next = self.head
            self.head = new_head
        else:
            self.head = new_head
            self.tail = new_head
    
    def add_last(self, val):
        new_tail = DequeNode(val)
        if self.tail:
            self.tail.next = new_tail
            new_tail.prev = self.tail 
            self.tail = new_tail
        else:
            self.head = new_tail
            self.tail = new_tail

    def pop_first(self):
        if self.head:
            saved_val = self.head.val
            new_head = self.head.next 
            new_head.prev = None
            self.head = new_head
            return saved_val
        else:
            raise IndexError("pop from empty deque")

    def pop_last(self):
        if self.tail:
            saved_val = self.tail.val
            new_tail = self.tail.prev
            new_tail.next = None 
            self.tail = new_tail
            return saved_val
        else:
            raise IndexError("pop form empty deque")

    def __len__(self):
        node = self.head
        length = 0
        while node:
            length += 1
            node = node.next 
        return length 

File: ts/src/data structures/deque.ts
import { DoubleLinkedListNode } from '../utils/linked_list_utility';
/**
 * This class represents the data structure known as a Deque,
 * implemented using doubly linked list
 * nodes.
 * @class @public
 */
class Deque<T> {
    head: null | DoubleLinkedListNode<T>;

    tail: null | DoubleLinkedListNode<T>;

    constructor() {
        this.head = null;
        this.tail = null;
    }

    /**
     * This method is used to return the first value stored inside of the
     * deque
     * @returns {T} Argument of type T, representing the first value inside
     * of the deque
     */
    peekFirst(): T {
        if (this.head) {
            return this.head.val;
        }

        throw Error('peek in empty deque');
    }

    /**
     * This method is used to return the last value stored inside of the
     * deque
     * @returns {T} Argument of type T, representing the last value inside
     * of the deque
     */
    peekLast(): T {
        if (this.tail) {
            return this.tail.val;
        }

        throw Error('peek in empty deque');
    }

    /**
     * This method is used to remove and return the first value stored
     * inside of the deque
     * @returns {T} Argument of type T, representing the first value inside
     * of the deque
     */
    popFirst(): T {
        if (this.head) {
            const retvalue = this.head.val;
            const newHead = this.head.next;
            if (newHead) {
                newHead.prev = null;
                this.head.next = null;
                this.head = newHead;
            } else {
                this.head = null;
                this.tail = null;
            }
            return retvalue;
        }

        throw Error('pop from empty deque');
    }

    popLast(): T {
        if (this.tail) {
            const retvalue = this.tail.val;
            const newTail = this.tail.prev;
            if (newTail) {
                newTail.next = null;
                this.tail.prev = null;
                this.tail = newTail;
            } else {
                this.head = null;
                this.tail = null;
            }
            return retvalue;
        }

        throw Error('pop from empty deque');
    }

    addFirst(value: T): void {
        const newHead = new DoubleLinkedListNode(value);
        if (this.head) {
            newHead.next = this.head;
            this.head.prev = newHead;
            this.head = newHead;
        } else {
            this.head = newHead;
            this.tail = newHead;
        }
    }

    addLast(value: T): void {
        const newTail = new DoubleLinkedListNode(value);
        if (this.tail) {
            newTail.prev = this.tail;
            this.tail.next = newTail;
            this.tail = newTail;
        } else {
            this.head = newTail;
            this.tail = newTail;
        }
    }

    length(): number {
        let length = 0;
        let node = this.head;
        while (node) {
            length += 1;
            node = node.next;
        }
        return length;
    }
}

export default Deque;
